spec/unit/configurer_spec.rb

```diff
@@ -1107,20 +1063,87 @@ def expects_neither_new_or_cached_catalog
 
-      Puppet::Node.indirection.terminus_class = :rest
-      Puppet::Resource::Catalog.indirection.terminus_class = :rest
+    describe "when the last used environment is available" do
+      let(:last_used_environment) { 'development' }
 
-      stub_request(:get, 'https://myserver:123/status/v1/simple/master').to_return(status: 200)
-      stub_request(:post, %r{https://myserver:123/puppet/v3/catalog}).to_return(status: 200)
-      node_request = stub_request(:get, %r{https://myserver:123/puppet/v3/node/}).to_return(status: 200)
+      before do
+        Puppet[:lastrunreport] = my_fixture('last_run_report.yaml')
+        Puppet[:lastrunfile] = file_containing('last_run_summary.yaml', <<~SUMMARY)
+        ---
+        version:
+          config: 1624882680
+          environment: development
+          puppet: 6.24.0
+        SUMMARY
+      end
 
-      configurer.run
+      it "prefers the environment set via cli" do
+        Puppet.settings.handlearg('--environment', 'usethis')
+        configurer.run
+
+        expect(configurer.environment).to eq('usethis')
+      end
+
+      it "prefers the environment set via config" do
+        FileUtils.mkdir_p(Puppet[:confdir])
+        set_puppet_conf(Puppet[:confdir], <<~CONF)
+        [main]
+        environment = usethis
+        CONF
+
+        Puppet.initialize_settings
+        configurer.run
+
+        expect(configurer.environment).to eq('usethis')
+      end
+
+      it "does not use the last used environment if given a catalog" do
+        configurer.run(catalog: catalog)
+
+        expect(configurer.environment).not_to be(last_used_environment)
+      end
 
-      expect(node_request).to have_been_requested.once
+      it "does not use the last used environment if use_cached_catalog = true" do
+        Puppet[:use_cached_catalog] = true
+        expects_cached_catalog_only(catalog)
+        configurer.run
+
+        expect(configurer.environment).not_to be(last_used_environment)
+      end
+
+      describe "when the environment is not configured" do
+        it "uses the environment found in lastrunfile if the key exists" do
+          configurer.run
+
+          expect(configurer.environment).to eq(last_used_environment)
+        end
+
+        it "falls back to the environment from lastrunreport if lastrunfile is not suitable" do
+          Puppet[:lastrunfile] = file_containing('last_run_summary.yaml', <<~SUMMARY)
+          ---
+          version:
+            config: 1624882680
+            puppet: 6.24.0
+          SUMMARY
+          configurer.run
+
+          expect(configurer.environment).to eq(last_used_environment)
+        end
+
+        it "uses the default environment if the last used one cannot be found" do
+          Puppet[:lastrunfile] = tmpfile('last_run_summary.yaml')
+          Puppet[:lastrunreport] = tmpfile('last_run_report.yaml')
+          configurer.run
+
+          expect(configurer.environment).not_to eq(last_used_environment)
```

Comment:
If neither contain an environment, do we fallback to `Puppet[:environment]/production` or do we send an empty environment as a query parameter, eg `/puppet/v3/catalog/:name?environment=&...`? Maybe good to expect a specific value here?